### DISCUSSION TITLE
Change Role?

### DISCUSSION THREAD

**Discussion by glowkeeper** [directus/directus#5037] [Archive]
Is it possible to change a user's role via the REST API?

**Comment by rijkvanzanten** [comment]
Yup! You can update the user and set the role field. See https://docs.directus.io/reference/api/system/users/#update-a-user

**Reply by glowkeeper** [reply]
Awesome - thanks :)

**Reply by glowkeeper** [reply]
I have that working, but I'm also trying to update via a custom URL, where I pass in `userid` and `role`. Something like this:
```
      const userid = req.body.userid;
      const userService = new UsersService({ schema: req.schema });
      await userService.update(req.body, userid);
```
...at least - that's my latest attempt - I've tried a few ways, all of which have failed, obvs :(

Er, help!

**Reply by RageBrahimi** [reply]
```javascript
const service = new UsersService({
  accountability: req.accountability,
  schema: req.schema,
});

const primaryKey = await service.update(req.body, req.params.pk);
```

You can actually see what going on [here](https://github.com/directus/directus/blob/237f709a5fa29da0302cb848d184282652656426/api/src/controllers/users.ts#L145).

**Reply by glowkeeper** [reply]
Actually, I think this was working, afterall:
```
     const userid = req.body.userid;
      const userService = new UsersService({ schema: req.schema });
      await userService.update(req.body, userid);
```

where I was calling it with:
```
{
    userid: "23453453451"
    role: "234523452345"
 }
 ```
 I _think_ the only difference to your working example is the stuff to do with _accountability: req.accountability_, which I don't really understand, in all honesty! And in all honesty, I should! Does that help define who can and can't run the update?

**Reply by glowkeeper** [reply]
So, just to continue this thread - I have another URL where I do this:
```
if ( !req.accountability.user ) {

        console.error(config.uRLs.resetPass.url + " unknown user");
        return next(new ServiceUnavailableException("Unavailable"));
    }
    
const userid = req.body.userid;
    const newPassword = {
      password: req.body.newPass
    };

    const userService = new UsersService({ schema: req.schema });
    userService
      .update(newPassword, userid)
```
What _I think_ you're showing me is that I can do the same thing by using _req.accountability_ directly in the query/update/whatever. i.e:
```
 const userService = new UsersService({ accountability: req.accountability, schema: req.schema });
    userService
      .update(newPassword, userid)
```
Right?

**Reply by RageBrahimi** [reply]
> Does that help define who can and can't run the update?

**Correct.**

A good practice is if you were to call a fresh custom endpoint **WITH** an `access_token` OR `Bearer TOKEN` then `console.log(req.accountability)` you will see it provides a `user` node.

Unless your custom endpoint publicly allows for updating user roles (obviously not recommended) then you'd want to ensure that the person who calls that endpoint is actually permitted to do so, hence the use of an `access_token`.

**Reply by glowkeeper** [reply]
Of course! I don't think that's what I was asking - I am aware that `access_token` comes bundled in _req.accountability_ - I _think_ I was asking how to use it more efficiently; i.e. does this:

```
const userService = new UsersService({ accountability: req.accountability, schema: req.schema });
const key = await userService.update(req.body, userid);
``` 
...achieve the same thing as the more verbose code below?
```
    if ( !req.accountability.user ) {
        return next(new ServiceUnavailableException("Unavailable"));
    }

    const userService = new UsersService({ schema: req.schema });
    const key = await userService.update(req.body, userid);
  ```

**Reply by glowkeeper** [reply]
(I'm aware all that needs wrapping in `try catch` etc. - just trying to understand conceptually...)

**Reply by rijkvanzanten** [reply]
@glowkeeper The first example actually kicks in the permissions checks for that particular user, so it's way more secure than the second example, which only checks if the user is logged in or not.

**Reply by glowkeeper** [reply]
@rijkvanzanten - Aha! Excellent - thanks!

**Reply by glowkeeper** [reply]
I think `accountability: req.accountability` has quickly become my favourite part of _Directus_ :)
